icu4j/maven-migration/toMaven.sh

```diff
@@ -0,0 +1,257 @@
+#!/usr/bin/env bash
+
+# Move the sources from main/classes and main/tests into a maven structure
+# - `main/classes/foo/src/java` go to `foo/src/main`, then split into `java` and `resources`
+# - `main/tests/foo/src/java` go to `foo/src/test`, then split into `java` and `resources`
+# - `main/tests/foo/src/META-INF` goes to ... `resources`
+#
+# From this:
+# classes/
+#   <module>/
+#     src/
+# test/
+#   <module>/
+#     src/
+#
+# To this (standard maven structure):
+# <module>/
+#   main/
+#     src/
+#       java/
+#       resources/
+#   test/
+#     src/
+#       java/
+#       resources/
+#
+# Then we cleanup older Eclipse project & launcher files, ant scripts, manifest.stub
+
+function rmDirIfExists() {
+  [ -d "$1" ] && rm -fr "$1"
+}
+function rmFileIfExists() {
+  [ -f "$1" ] && rm "$1"
+}
+function safeMoveDir() {
+  export FOLDER_NAME=$1
+  export FROM_FOLDER=$2
+  export TO_FOLDER=$3
+
+  if [ ! -d "$FROM_FOLDER/$FOLDER_NAME" ]; then
+    echo "  No '$FROM_FOLDER/$FOLDER_NAME' to move."
+    return
+  fi
+  if [ -e "$TO_FOLDER/$FOLDER_NAME" ]; then
+    echo "  Error: folder '$TO_FOLDER/$FOLDER_NAME' elready exists!"
+    exit
+  else
+    mkdir -p $TO_FOLDER
+    mv $FROM_FOLDER/$FOLDER_NAME $TO_FOLDER/
+    echo "  Moving '$FOLDER_NAME' folder from '$FROM_FOLDER' to '$TO_FOLDER'"
+  fi
+}
+
+# Split the content of the `java` foldert into `java` and `resources`, the way maven wants it
+function splitJavaToResources() {
+  # Should point to the folder containing `java`, either main/ or test/
+  export BASE_FOLDER=$1
+
+  echo "  Splitting '$BASE_FOLDER/java/' into '$BASE_FOLDER/java/' and '$BASE_FOLDER/resources/'"
+  # copy `java` to `resources`
+  cp -R $BASE_FOLDER/java/ $BASE_FOLDER/resources/
+  # delete all not `.java` from `java`
+  find $BASE_FOLDER/java/      -type f -not -name *.java -exec rm {} \;
+  # delete all `.java` from `resources`
+  find $BASE_FOLDER/resources/ -type f      -name *.java -exec rm {} \;
+}
+
+function removeEclipseProjectFiles() {
+  # Should point to the old folder (to be move), containing the eclipse / ant files
+  export BASE_FOLDER=$1
+
+  # Cleanup Eclipse project & launcher files, ant projects, other stuff
+  # Eclipse
+  rmDirIfExists  "$BASE_FOLDER/.externalToolBuilders/"
+  rmDirIfExists  "$BASE_FOLDER/.settings/"
+  rmFileIfExists "$BASE_FOLDER/.project"
+  rmFileIfExists "$BASE_FOLDER/.classpath"
+  # Ant
+  rmFileIfExists "$BASE_FOLDER/build.properties"
+  rmFileIfExists "$BASE_FOLDER/build.xml"
+  rmFileIfExists "$BASE_FOLDER/manifest.stub"
+  rmFileIfExists "$BASE_FOLDER/findbugs-exclude.xml"
+  find $BASE_FOLDER/ -type f -name '*.launch' -exec rm {} \;
+}
+
+function removeEmptyFolders() {
+  export BASE_FOLDER=$1
+  for ((n = 0; n < 15; n++)) do
```

Comment:
Why do we need a loop of a seemingly arbitrary number of iterations?

Reply:
I try to remove the empty folders.
These would be left-overs from moving files / folders out of them.

But I can't remove recursive because it also removes the files.

Let's say I have the empty folder `main/classes/core/src/`
I can't delete `main/` because it is not empty.
And I can't force delete, because it might contain files (should not, but I want to detect that).

So by doing this loop on `main/classes/core/src/` what happens is:

loop 1:
```
rm `main/classes/` => FAILS, not empty
rm `main/classes/core/` => FAILS, not empty
rm `main/classes/core/src/` => OK
```
loop 2:
```
rm `main/classes/` => FAILS, not empty
rm `main/classes/core/` => OK
```
loop 3:
```
rm `main/classes/` => OK
```

I agree that 15 steps might be a bit exaggerated :-)

---

I'm open to other ways to do this, but I was unable to find anything using most common tools that one can find on any machine.
I tried to stay away from fancy tools that need installing, or python scripts.
It is an one-off operation.

Reply:
A comment on the loop with a brief version of this explanation would be nice.

Reply:
Done.